Fix #44443: DAGs state change from running to failed, when paused, to avoid the increment of their duration

PNL0 opened the pull request:
The bug happened because the API function used to patch a DAG, only updated the is_paused field in the DagModel structure. This made the state of a running DAG to stay on running and consequently the duration will keep on increasing even though the DAG is paused.

To fix this, I added some logic to the API function to also update the DAG state, depending on the value of is_paused field (paused makes the state change from running to failed).

Closes: #44443

bugraoz93 commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -235,6 +236,23 @@ def patch_dag(
     data = patch_body.model_dump(include=fields_to_update, by_alias=True)
 
+    is_paused = data.get("is_paused")
+    if is_paused is not None:
```
```suggestion
    if is_paused:
```

PNL0 replied:
I think I should remove the if-clause, because is_paused is a boolean that will always be true or false (since it represents the value of is_paused field from the DagModel, after the update, it's impossible to be None).

bugraoz93 commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -235,6 +236,23 @@ def patch_dag(
+        ).all()
+
+        for dag_run in active_dag_runs:
+            if is_paused:
```
I think there is a problem here. It cannot get into else here since the parent `if` is already ensuring it isn't `None`.

PNL0 replied:
> I think there is a problem here. It cannot get into else here since the parent `if` is already ensuring it isn't `None`.

My idea, with the if-else clause, is to cover the cases when the user pauses the dag (is_paused = true) and unpauses the dag (is_paused = false, so the else clause). 

Maybe this became confusing because of the parent if-clause (to check the None), but has I said in the other comment, I think that I can safely remove it.

bugraoz93 reviewed (commented):
I am not sure if this follows this comment in the issue stated in the MR title.

https://github.com/apache/airflow/issues/44443#issuecomment-2513833985

>> I'm just into something, if we no need to rush, I will happily raise PR

>Raising a PR to fix the cluster activity (Top 5 running dags) is step forward. Maybe the issue in the title isn't matching with the PR. Could you please also update the description and include `closes:#<issue_number>` or `relates:#<issue_number>`?

ephraimbuddy commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -246,6 +247,23 @@ def patch_dag(
+
+    for dag_run in active_dag_runs:
+        if is_paused:
+            dag_run.set_state(DagRunState.QUEUED)
```
I disagree with this. Why will a dag run go from running to queued state?

PNL0 replied:
> I disagree with this. Why will a dag run go from running to queued state?

I think that if a DAG is paused, then the dag run shouldn't continue to run (which implies a change in its state from running to other state that keeps the dag run paused). 

I chose queued state, because a similar behavior already happens in Airflow, when you trigger a paused dag (dag run state goes to queued).

ephraimbuddy replied:
> > I disagree with this. Why will a dag run go from running to queued state?
> 
> I think that if a DAG is paused, then the dag run shouldn't continue to run (which implies a change in its state from running to other state that keeps the dag run paused).
> 
> I chose queued state, because a similar behavior already happens in Airflow, when you trigger a paused dag (dag run state goes to queued).

What about if you pause a running DAG? Maybe you can check the behaviour in Airflow 2 UI

PNL0 replied:
> What about if you pause a running DAG? Maybe you can check the behaviour in Airflow 2 UI

If you pause a running dag, then the dag run state will be stuck on running and as result the duration keeps on increasing permanently, which it's not the desired behavior for pause functionality. 

This behavior happens in Airflow 2 and 3 UIs (also if I'm not mistaken, the issue was reported, when airflow 3 UI wasn't released).

That's why I think the dag run state can't stay on running when the DAG is paused.

ephraimbuddy replied:
What do you think about failing all running dagruns once the dag is paused? Skipping those that are not running. Just the way we handle dagrun timeout

PNL0 replied:
> What do you think about failing all running dagruns once the dag is paused? Skipping those that are not running. Just the way we handle dagrun timeout

I think it's a good idea. 
I tried to implement what you described in my updated commit.

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -246,6 +247,23 @@ def patch_dag(
+        else:
+            if dag_run.state == DagRunState.QUEUED:
+                dag_run.set_state(DagRunState.RUNNING)
+
```
I don't think you need that. The scheduler will automatically Run this dagrun if it is unpaused.
(`_start_queued_dagruns`, `get_queued_dag_runs_to_set_running`)

PNL0 replied:
I agree. 
In the update, I removed that code, since it wasn't necessary, as you said.

pierrejeambrun reviewed (commented):
I feel like we should make the behavior consistent. There are other places where we set the dag_run to paused and it should also set the state to QUEUED there.

Also this means that the dagrun will remain forever in the queued state ? I'm not sure it's great.

Maybe we should mark the dag_run as failed (fail the current running tasks and let the run settle -> other tasks will be skipped for failed upstream, and the dag_run state will settle to failed) ?

cc: @potiuk All that sounds like an impactful change and maybe we should discuss that on the mailing list to see what others think (behavior + technical implication)

PNL0 committed:
Fix apache#44443: Paused DAGs stuck on running.

If dags are paused after they start running their duration keeps on
increasing, because their state doesn't change (remains running).
To fix this issue, now when a dag is paused, the state switches from
running to failed and the same happens for every running task in the
dag run.

ephraimbuddy commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -246,6 +248,29 @@ def patch_dag(
+    for dag_run in active_dag_runs:
+        if is_paused:
+            dag_run.set_state(DagRunState.FAILED)
+            running_task_instances = session.scalars(
+                select(TaskInstance).filter(
+                    TaskInstance.dag_id == dag_run.dag_id,
+                    TaskInstance.run_id == dag_run.run_id,
+                    TaskInstance.state == TaskInstanceState.RUNNING,
+                )
+            ).all()
+            for task_instance in running_task_instances:
+                task_instance.set_state(TaskInstanceState.FAILED)
```
As @pierrejeambrun suggested, we should discuss this change in behaviour in dev list.

pierrejeambrun replied:
Dev mailing list email address is `dev@airflow.apache.org`. And you can take a look the archives there to see how things are discussed:
https://lists.apache.org/list.html?dev@airflow.apache.org

PNL0 replied:
Hello @ephraimbuddy @pierrejeambrun,

I created a post in the dev mailing list, as you suggested.

pierrejeambrun reviewed (commented):
Yes, following the discussion on the mailing list, this is expected and there is nothing to fix for now since the `top 5 longest dagrun` are not yet implemented in AF3. We will keep that in mind if this feature were to come back.
